pkg/base/capacityplanner/node_capacity.go

```diff
@@ -17,172 +17,150 @@
 	"github.com/dell/csi-baremetal/pkg/base/util"
 )
 
-type nodeCapacity struct {
-	// AC cache
-	capacity ACMap
-	// store original versions of modified ACs
-	origAC ACMap
+type reservedCapacity struct {
+	Size         int64
+	StorageClass string
 }
 
-// registerAC register AC in internal cache
-func (nc *nodeCapacity) registerAC(ac *accrd.AvailableCapacity) {
-	nc.capacity[ac.Name] = ac
-}
+type reservedACs map[string]*reservedCapacity
 
-func (nc *nodeCapacity) saveOriginalAC(ac *accrd.AvailableCapacity) {
-	acInCache, ok := nc.capacity[ac.Name]
-	if !ok {
-		return
-	}
-	if nc.origAC == nil {
-		nc.origAC = ACMap{}
-	}
-	// we need to save original AC only once
-	if _, ok := nc.origAC[ac.Name]; !ok {
-		nc.origAC[ac.Name] = acInCache.DeepCopy()
-	}
-}
+type scToACOrder map[string][]string
 
-func (nc *nodeCapacity) getOriginalAC(acName string) *accrd.AvailableCapacity {
-	if ac, ok := nc.origAC[acName]; ok {
-		return ac
-	}
-	if ac, ok := nc.capacity[acName]; ok {
-		return ac
-	}
-	return nil
-}
-
-// removeAC remove AC from internal cache
-func (nc *nodeCapacity) removeAC(ac *accrd.AvailableCapacity) {
-	delete(nc.capacity, ac.Name)
+type nodeCapacity struct {
+	node string
+	// ACs for the related node
+	acs ACMap
+	// AC Name: {AC Size, AC Storage Class}
+	reservedACs reservedACs
+	// Storage Class Name: the list of AC names, which can be selected
+	acsOrder scToACOrder
 }
 
-// selectACForVolume select AC for volume
-// will modify nodeCapacity AC cache
-func (nc *nodeCapacity) selectACForVolume(vol *genV1.Volume) *accrd.AvailableCapacity {
-	if util.IsStorageClassLVG(vol.StorageClass) {
-		return nc.selectACForLVMVolume(vol)
+func newNodeCapacity(node string, acs []accrd.AvailableCapacity, acrs []acrcrd.AvailableCapacityReservation) *nodeCapacity {
+	// Check for no capacity on node
+	if len(acs) == 0 {
+		return nil
 	}
-	return nc.selectACForFullDriveVolume(vol)
-}
 
-// selectACForFullDriveVolume selects AC for ANY SC or for other "full drive" storage classes.
-func (nc *nodeCapacity) selectACForFullDriveVolume(vol *genV1.Volume) *accrd.AvailableCapacity {
-	scToACMap := nc.getStorageClassToACMapping()
+	// Chose AC only for the related node
+	acs = FilterACList(acs, func(ac accrd.AvailableCapacity) bool {
+		return ac.Spec.NodeId == node
+	})
 
-	if len(scToACMap[vol.StorageClass]) == 0 &&
-		vol.StorageClass != v1.StorageClassAny {
-		return nil
-	}
-	requiredSize := vol.GetSize()
-	// filter out non relevant storage classes
-	filteredMap := SCToACMap{}
-	if vol.StorageClass == v1.StorageClassAny {
-		for sc, acs := range scToACMap {
-			// for any SC we need to check for non LVG only
-			if !util.IsStorageClassLVG(sc) {
-				// TODO Take into account drive technology for SC ANY https://github.com/dell/csi-baremetal/issues/231
-				// map must be sorted HDD->SSD->NVMe
-				filteredMap[sc] = acs
-			}
+	// Sort AC to have the persistent order for each reservation
+	sort.Slice(acs, func(i, j int) bool {
```

Comment:
How will this sort method work on a big scale in term of the speed and memory consumption?

Reply:
It requires <80ms for 10000 values for AC structs. I guess it brings small impact to provision process. I have no info about memory usage, but can check